### PR TITLE
Add opensearch sink security

### PR DESCRIPTION
**Pull request by lizsnyder**
### Description

Adds security documentation for Data Prepper opensearch sink plugin.

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Naarcha-AWS — APPROVED** [review]

**Comment by Naarcha-AWS** [comment]
Pushing this through to editorial in lieu of waiting for another Doc review.

**Review by ariamarble — APPROVED** [review]

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 1]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
+- Index: `.opendistro-ism-config`; Index permission: `indices_all`
+- Index: `*`; Index permission: `manage_aliases`
+
+For instructions to map users to roles, see [Map users to roles]({{site.url}}{{site.baseurl}}/security/access-control/users-roles/#map-users-to-roles).
+
+## Amazon OpenSearch Service domain security
+
+The OpenSearch sink plugin can send data to an [Amazon OpenSearch Service](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/what-is.html) domain, which uses Identity and Access Management (IAM) for security. The plugin uses the default credential chain. Run `aws configure` using the [AWS CLI](https://aws.amazon.com/cli/) to set your credentials.
+
+Make sure the credentials that you configure have the required IAM permissions. The following domain access policy demonstrates the minimum required permissions for the sink to work:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:108]:
```suggestion
Make sure the credentials that you configure have the required IAM permissions. The following domain access policy demonstrates the minimum required permissions:
```

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 2]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
+      },
+      "Action": "es:ESHttpGet",
+      "Resource": "arn:aws:es:us-east-1:<AccountId>:domain/<domain-name>/_cluster/settings"
+    }
+  ]
+}
+```
+
+For instructions to configure the domain access policy, see [Resource-based policies
+](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html#ac-types-resource) in the Amazon OpenSearch Service documentation.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:141]: Is this line break intentional?

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 3]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
+
+### Fine-grained access control
+
+If your OpenSearch Service domain uses [fine-grained access control
+](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac.html), the OpenSearch sink plugin requires some additional configuration.
+
+#### IAM ARN as master user
+
+If you're using an IAM ARN as the master user, include the `aws_sigv4` option in your sink configuration:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:150]:
```suggestion
If you're using an IAM Amazon Resource Name (ARN) as the master user, include the `aws_sigv4` option in your sink configuration:
```

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 4]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
+
+```yaml
+...
+sink:
+    opensearch:
+      hosts: ["https://your-fgac-amazon-opensearch-service-endpoint"]
+      aws_sigv4: true
+```
+
+Run `aws configure` using the AWS CLI to set your credentials to the master IAM user. If you don't want to use the master user, you can specify a different IAM role using the `aws_sts_role_arn` option. The plugin will then use this role to sign requests to the domain sink. The ARN that you specify must be included in the [domain access policy]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/#amazon-opensearch-service-domain-security).
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:160]:
```suggestion
Run `aws configure` using the AWS CLI to set your credentials to the master IAM user. If you don't want to use the master user, you can specify a different IAM role using the `aws_sts_role_arn` option. The plugin will then use this role to sign requests sent to the domain sink. The ARN that you specify must be included in the [domain access policy]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/#amazon-opensearch-service-domain-security).
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:160]: In the first sentence, do we mean something like "Run `aws configure` using the AWS CLI to use the master IAM user credentials"?

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 5]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
+
+For instructions to configure the domain access policy, see [Resource-based policies
+](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html#ac-types-resource) in the Amazon OpenSearch Service documentation.
+
+### Fine-grained access control
+
+If your OpenSearch Service domain uses [fine-grained access control
+](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac.html), the OpenSearch sink plugin requires some additional configuration.
+
+#### IAM ARN as master user
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:148]: Global: Is "master user" the accepted term, or should it be something like "primary user"?

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 6]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
+ ]
+  ```
+
+  ***Important***: Make sure to replace the Amazon Resource Name (ARN) in the `Principal` element with the ARN of the pipeline role that you created in the previous section.
+
+  For instructions to create a collections, see [Creating collections](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/serverless-manage.html#serverless-create) in the Amazon OpenSearch Service documentation.
+
+### Create a pipeline
+
+Within your `pipelines.yaml` file, specify the OpenSearch Serverless collection endpoint in the `hosts` option. In addition, you must set the `serverless` option to `true`. Specify the pipeline role in the `sts_role_arn` option.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/sinks/opensearch.md:262]:
```suggestion
Within your `pipelines.yaml` file, specify the OpenSearch Serverless collection endpoint as the `hosts` option. In addition, you must set the `serverless` option to `true`. Specify the pipeline role in the `sts_role_arn` option:
```

**Review by natebower — COMMENTED** [review]
@Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

File: _data-prepper/pipelines/configuration/sinks/opensearch.md [hunk 7]
```diff
@@ -29,16 +67,219 @@ index | Conditionally | String | Name of the export index. Applicable and requir
...
 number_of_replicas | No | Integer | The number of replica shards each primary shard should have on the destination OpenSearch server. For example, if you have 4 primary shards and set number_of_replicas to 3, the index has 12 replica shards. This parameter is effective only when `template_file` is either explicitly provided in Sink configuration or built-in. If this parameter is set, it would override the value in index template file. For more information, see [Create index]({{site.url}}{{site.baseurl}}/api-reference/index-apis/create-index/).
 
-<!--- ## Configuration
+## OpenSearch cluster security
+
+In order to send data to an OpenSearch cluster using the OpenSearch sink plugin, you must specify your username and password within the pipeline configuration. The following example `pipelines.yaml` file demonstrates how to specify Admin security credentials:
```
[the remark this answers is not in the document]
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sinks/opensearch.md:78]:
```suggestion
In order to send data to an OpenSearch cluster using the `opensearch` sink plugin, you must specify your username and password within the pipeline configuration. The following example `pipelines.yaml` file demonstrates how to specify Admin security credentials:
```

[the remark this answers is not in the document]
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sinks/opensearch.md]: Addressed this in previous comment about `opensearch` sink.

File: _data-prepper/pipelines/configuration/sinks/opensearch.md (hunk 5 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sinks/opensearch.md:148]: Will switch to primary.

Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sinks/opensearch.md:148]: Amending my previous comment. IAM still uses the term `master` when referring to the role of primary user (https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac-walkthrough-iam.html) Will hold off on switching this to remain consistent with the IAM offering.

File: _data-prepper/pipelines/configuration/sinks/opensearch.md (hunk 2 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sinks/opensearch.md:141]: Fixed.

File: _data-prepper/pipelines/configuration/sinks/opensearch.md (hunk 4 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sinks/opensearch.md]: Modified.

**Pull request merged**